arbnode/maintenance.go

```diff
@@ -136,41 +141,102 @@ func wentPastTimeOfDay(before time.Time, after time.Time, timeOfDay int) bool {
 	return prevMinutes < dbCompactionMinutes && newMinutes >= dbCompactionMinutes
 }
 
-func (mr *MaintenanceRunner) maybeRunMaintenance(ctx context.Context) time.Duration {
+func (mr *MaintenanceRunner) GetLastMaintenanceTime() (time.Time, error) {
+	milli := mr.lastMaintenance.Load()
+	if milli == 0 {
+		return time.Time{}, errors.New("maintenance running")
+	}
+	return time.UnixMilli(milli), nil
+}
+
+func (mr *MaintenanceRunner) setMaintenanceDone() {
+	milli := time.Now().UnixMilli()
+	prev := mr.lastMaintenance.Swap(milli)
+	if prev != 0 {
+		log.Warn("warning: maintenance ran in parallel", "current", time.UnixMilli(milli), "prev", time.UnixMilli(prev))
+	}
+}
+
+func (mr *MaintenanceRunner) setMaintenanceStart() error {
+	prev := mr.lastMaintenance.Swap(0)
+	if prev == 0 {
+		return errors.New("already running")
+	}
+	return nil
+}
+
+func (mr *MaintenanceRunner) maybeRunScheduledMaintenance(ctx context.Context) time.Duration {
 	config := mr.config()
 	if !config.enabled {
 		return time.Minute
 	}
 
 	now := time.Now().UTC()
 
-	if !wentPastTimeOfDay(mr.lastMaintenance, now, config.minutesAfterMidnight) {
+	lastMaintenance, err := mr.GetLastMaintenanceTime()
+	if err != nil {
 		return time.Minute
 	}
 
-	if mr.seqCoordinator == nil {
-		mr.lastMaintenance = now
-		mr.runMaintenance()
+	if !wentPastTimeOfDay(lastMaintenance, now, config.minutesAfterMidnight) {
 		return time.Minute
 	}
 
+	err = mr.attemptMaintenance(ctx)
+	if err != nil {
+		log.Warn("scheduled maintenance error", "err", err)
+	}
+
+	return time.Minute
+}
+
+func (mr *MaintenanceRunner) Trigger() error {
+	if !mr.config().Triggerable {
+		return errors.New("maintenance not configured to be triggerable")
+	}
+	// error if already running
```

Comment:
I wouldn't return an error for the API call, I would return a message, or at least a non error status code, indicating that maintenance is already running

Reply:
returned nil for now